src/oauth/sessionManager.ts

```diff
@@ -0,0 +1,808 @@
+import { type AxiosInstance } from "axios";
+import * as vscode from "vscode";
+
+import { CoderApi } from "../api/coderApi";
+import { type ServiceContainer } from "../core/container";
+import { type Deployment } from "../core/deployment";
+import { type LoginCoordinator } from "../login/loginCoordinator";
+
+import { OAuthMetadataClient } from "./metadataClient";
+import {
+	CALLBACK_PATH,
+	generatePKCE,
+	generateState,
+	toUrlSearchParams,
+} from "./utils";
+
+import type { SecretsManager, StoredOAuthTokens } from "../core/secretsManager";
+import type { Logger } from "../logging/logger";
+
+import type { OAuthError } from "./errors";
+import type {
+	ClientRegistrationRequest,
+	ClientRegistrationResponse,
+	OAuthServerMetadata,
+	RefreshTokenRequestParams,
+	TokenRequestParams,
+	TokenResponse,
+	TokenRevocationRequest,
+} from "./types";
+
+const AUTH_GRANT_TYPE = "authorization_code" as const;
+const REFRESH_GRANT_TYPE = "refresh_token" as const;
+const RESPONSE_TYPE = "code" as const;
+const PKCE_CHALLENGE_METHOD = "S256" as const;
+
+/**
+ * Token refresh threshold: refresh when token expires in less than this time.
+ */
+const TOKEN_REFRESH_THRESHOLD_MS = 10 * 60 * 1000;
+
+/**
+ * Default expiry time for OAuth access tokens when the server doesn't provide one.
+ */
+const ACCESS_TOKEN_DEFAULT_EXPIRY_MS = 60 * 60 * 1000;
+
+/**
+ * Minimum time between refresh attempts to prevent thrashing.
+ */
+const REFRESH_THROTTLE_MS = 30 * 1000;
+
+/**
+ * Background token refresh check interval.
+ */
+const BACKGROUND_REFRESH_INTERVAL_MS = 5 * 60 * 1000;
+
+/**
+ * Minimal scopes required by the VS Code extension.
+ */
+const DEFAULT_OAUTH_SCOPES = [
+	"workspace:read",
+	"workspace:update",
+	"workspace:start",
+	"workspace:ssh",
+	"workspace:application_connect",
+	"template:read",
+	"user:read_personal",
+].join(" ");
```

Comment:
Did you have a way to test if these scopes were sufficient?  Not sure if there is a way right now, maybe if we try to generate a key manually with these scopes (I think this can be done on the account page) and paste it in to make sure everything works.